src/js/common/components/RequiredLoginView.jsx

```diff
@@ -8,111 +8,113 @@ import LoadingIndicator from '../../common/components/LoadingIndicator';
 import Main from '../../login/containers/Main';
 
 const healthTools = ['health-records', 'rx', 'messaging'];
+const nextQuery = { next: window.location.pathname };
+const signInUrl = appendQuery('/', nextQuery);
 
 class RequiredLoginView extends React.Component {
-  isServiceAvailable() {
-    const requiredApp = this.props.serviceRequired;
-    const userServices = this.props.userProfile.services;
-    // Checks that
-    // 1) session has a valid authentication token, and
-    // 2) the application being loaded (requiredApp) is in the list of
-    //    applications/services the user is authorized to use (userServices)
-    // TODO: replace indexOf() once NodeJS versions in all environments support includes()
-
-    if (sessionStorage.userToken && userServices) {
-      if (Array.isArray(requiredApp)) {
-        return intersection(userServices, requiredApp).length > 0;
-      }
-      return userServices.indexOf(requiredApp) !== -1;
+  componentDidUpdate() {
+    const isLoading = this.props.user.profile.loading;
+    if (!isLoading && this.shouldSignIn()) {
+      window.location.replace(signInUrl);
     }
+  }
+
+  shouldSignIn = () => !this.props.user.login.currentlyLoggedIn;
+
+  // Checks that (1) session has a valid authentication token and
+  // (2) the user is authorized to use services required by this application
+  isAccessible = () => {
+    const { serviceRequired, user } = this.props;
+    const userServices = user.profile.services;
+    const hasRequiredServices = userServices && (
+      Array.isArray(serviceRequired) ?
+        intersection(userServices, serviceRequired).length > 0 :
+        userServices.includes(serviceRequired)
+    );
 
-    return false;
+    return sessionStorage.userToken && hasRequiredServices;
   }
 
-  renderContent() {
-    const { userProfile, serviceRequired: requiredApp } = this.props;
+  renderVerifiedContent = () => {
+    const { serviceRequired, user } = this.props;
+
+    // Ignore required verification when accessing health tools with MHV sign-in.
+    const hasMhvAccess =
+      healthTools.includes(serviceRequired) &&
+      user.profile.authnContext === 'myhealthevet';
+
+    if (user.profile.verified || hasMhvAccess) {
+      // TODO: Delete the logic around attemptingAppealsAccess once we
+      // resolve the MVI/Appeals Users issues.
+      // if app we are trying to access includes appeals,
+      // bypass the checks for user profile status
+      const attemptingAppealsAccess = Array.isArray(serviceRequired) ?
+        serviceRequired.includes('appeals-status') :
+        serviceRequired === 'appeals-status';
+
+      if (user.profile.status === 'SERVER_ERROR' && !attemptingAppealsAccess) {
```

Comment:
If the availability of certain services are predicated on being able to access MVI data, perhaps we shouldn't even be checking this here. Ideally, if the API response can be granular enough to inform us that service X is unavailable for Y reason, the F/E won't have to gather separate pieces of data and interpret why the user can't access this service (whether it's right now or ever). @saneshark

Reply:
I agree. I'm not a huge fan of doing this in the login view component. Should be based on available services.